dmoj/judge.py

```diff
@@ -459,69 +491,132 @@ def _grade_cases(self) -> Generator[Tuple[IPC, tuple], None, None]:
+            test_iter = iter(tests)
+            for test in test_iter:
+                if not should_run(test, failed_tests):
+                    failed = True
+                    yield from skip_toplevel(test)
+                elif isinstance(test, Batch):
+                    yield IPC.BATCH_BEGIN, (test.batch,)
+
+                    batch_verdict = yield from grade_batched_cases(test.batched_cases)
+                    failed = batch_verdict != Result.AC
+
+                    yield IPC.BATCH_END, (test.batch,)
                 else:
-                    result = self.grader.grade(case)
+                    result = grade(test)
+                    failed = result.result_flag & Result.WA
+                    yield IPC.RESULT, (None, test.position, result)
+
+                if failed:
+                    failed_tests.add(test.config.raw_config_id)
+                    if not test.points or self.submission.short_circuit:
+                        for remaining_test in test_iter:
+                            failed_tests.add(remaining_test.config.raw_config_id)
+                            yield from skip_toplevel(remaining_test)
+
+                        return failed_tests, GradeManyResult.SKIP_ALL
+
+            return failed_tests, GradeManyResult.OK
 
-                    # If the submission was killed due to a user-initiated abort, any result is meaningless.
-                    if self._abort_requested:
-                        yield IPC.GRADING_ABORTED, ()
-                        return
+        def should_run_pretest(case: TopLevelCase, failed_so_far: Set[int]) -> bool:
+            if case.dependencies is None:
+                # Default: depends on nothing.
+                return True
 
-                    if result.result_flag & Result.WA:
-                        # If we failed a 0-point case, we will short-circuit every case after this.
-                        is_short_circuiting_enabled |= not case.points
+            if case.dependencies & failed_so_far:
+                return False
 
-                        # Short-circuit if we just failed a case in a batch, or if short-circuiting is currently enabled
-                        # for all test cases (either this was requested by the site, or we failed a 0-point case in the
-                        # past).
-                        is_short_circuiting |= batch_number is not None or is_short_circuiting_enabled
+            return True
 
-                # Legacy hack: we need to allow graders to read and write `proc_output` on the `Result` object, but the
-                # judge controller only cares about the trimmed output, and shouldn't waste memory buffering the full
-                # output. So, we trim it here so we don't run out of memory in the controller.
-                result.proc_output = result.output
-                yield IPC.RESULT, (batch_number, case_number, result)
+        failed_pretests: Set[int] = set()
```

Comment:
Rather than have this state here captured by the closure in `should_run_main_test`, what if we took it as the first parameter to `should_run_main_test`, and then applied it with `partial(...`?